overlord/devicestate/handlers.go

```diff
@@ -698,3 +703,127 @@ func fetchKeys(st *state.State, keyID string) (errAcctKey error, err error) {
 		keyID = a.SignKeyID()
 	}
 }
+
+func snapState(st *state.State, name string) (*snapstate.SnapState, error) {
+	var snapst snapstate.SnapState
+	err := snapstate.Get(st, name, &snapst)
+	if err != nil && err != state.ErrNoState {
+		return nil, err
+	}
+	return &snapst, nil
+}
+
+func makeRollbackDir(name string) (string, error) {
+	snapRollbackDir := filepath.Join(dirs.SnapRollbackDir, name)
+
+	if err := os.MkdirAll(dirs.SnapRollbackDir, 0755); err != nil {
+		return "", err
+	}
+
+	if err := os.Mkdir(snapRollbackDir, 0755); err != nil && !os.IsExist(err) {
+		return "", err
+	}
+
+	return snapRollbackDir, nil
+}
+
+func currentGadgetInfo(snapst *snapstate.SnapState) (*gadget.Info, error) {
+	var gi *gadget.Info
+
+	currentInfo, err := snapst.CurrentInfo()
+	if err != nil && err != snapstate.ErrNoCurrent {
+		return nil, err
+	}
+	if currentInfo != nil {
+		gi, err = snap.ReadGadgetInfo(currentInfo, release.OnClassic)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return gi, nil
+}
+
+func pendingGadgetInfo(snapsup *snapstate.SnapSetup) (*gadget.Info, error) {
+	info, err := snap.ReadInfo(snapsup.InstanceName(), snapsup.SideInfo)
+	if err != nil {
+		return nil, err
+	}
+	update, err := snap.ReadGadgetInfo(info, release.OnClassic)
+	if err != nil {
+		return nil, err
+	}
+	return update, nil
+}
+
+func gadgetCurrentAndUpdate(st *state.State, snapsup *snapstate.SnapSetup) (current *gadget.Info, update *gadget.Info, err error) {
+	snapst, err := snapState(st, snapsup.InstanceName())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	currentInfo, err := currentGadgetInfo(snapst)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	newInfo, err := pendingGadgetInfo(snapsup)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return currentInfo, newInfo, nil
+}
+
+var (
+	gadgetUpdate = nopGadgetOp
+)
+
+func nopGadgetOp(current, update *gadget.Info, rollbackRootDir string) error {
+	return nil
+}
+
+func (m *DeviceManager) doUpdateGadget(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	snapsup, err := snapstate.TaskSnapSetup(t)
+	if err != nil {
+		return err
+	}
+
+	current, update, err := gadgetCurrentAndUpdate(t.State(), snapsup)
+	if err != nil {
+		return err
+	}
+	if current == nil {
+		// XXX: no updates during first boot & seeding
+		return nil
+	}
+
+	snapRollbackDir, err := makeRollbackDir(snapsup.InstanceName())
+	if err != nil {
+		return fmt.Errorf("cannot prepare update rollback directory: %v", err)
+	}
+
+	st.Unlock()
+	err = gadgetUpdate(current, update, snapRollbackDir)
+	st.Lock()
+	if err != nil {
+		if err == gadget.ErrNoUpdate {
+			// no update needed
+			return nil
+		}
+		return err
+	}
+
+	t.SetStatus(state.DoneStatus)
+
+	if err := os.RemoveAll(snapRollbackDir); err != nil && !os.IsNotExist(err) {
```

Comment:
what's the plan for the rolllback dir?  use them for undo?  or keep them around and give the user a way to invoke them?

what should "snap revert gadget" do with them?  should the name contain also the original revision?

Reply:
The plan was to keep it for undo and remove in cleanup. Then we revised the ideas about undo, and the rollback dir ended up being used for reverting up after a failed update attempt. Where a failed update means `gadget.Update` errors internally and tries to restore the original state, no reboot occurs.

At this point I'm thinking how far we can actually go with the cleanup and whether there should be limits on what can be restored. The main issue I see is storage space on the host and a bit vague behavior of filesystem updates, we we'd need to copy out an unpredictable number of files (or a whole tree).

Reply:
I thought a bit more about the issue of "internal" undos in the tasks,  for debugging and for supporting that probably the best is for the actual backup/update to create  a not human unfriendly but machine readable append only log file in the rollback there probably.  (I wouldn't use the state because of the churn and because the changes done here are mostly orthogonal to the state of the gadget as  a snap or other snaps).

Reply:
The idea of some log in the rollback dir seems still good but can be done when we do write gadgetUpdate for real?

Reply:
Yes, adding a log then wold be preferred by me too. Some initial version of update code is already for review, but can be easily extended further to support a log of changes done in backup/deploy/rollback passes.